### ISSUE TITLE
How can I export a YOLOE ONNX model with a fused BN layer?

### ISSUE THREAD

**Issue by xifeng0126** [ultralytics/ultralytics#21266]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I noticed that the Yoloe case is not considered in the ` fuse()` method, so when exporting the Yoloe model using `export()`, the model will not be fused. How should I implement it?
```
   def fuse(self, verbose=True):
        """
        Fuse the `Conv2d()` and `BatchNorm2d()` layers of the model into a single layer for improved computation
        efficiency.

        Returns:
            (torch.nn.Module): The fused model is returned.
        """
        if not self.is_fused():
            for m in self.model.modules():
                if isinstance(m, (Conv, Conv2, DWConv)) and hasattr(m, "bn"):
                    if isinstance(m, Conv2):
                        m.fuse_convs()
                    m.conv = fuse_conv_and_bn(m.conv, m.bn)  # update conv
                    delattr(m, "bn")  # remove batchnorm
                    m.forward = m.forward_fuse  # update forward
                if isinstance(m, ConvTranspose) and hasattr(m, "bn"):
                    m.conv_transpose = fuse_deconv_and_bn(m.conv_transpose, m.bn)
                    delattr(m, "bn")  # remove batchnorm
                    m.forward = m.forward_fuse  # update forward
                if isinstance(m, RepConv):
                    m.fuse_convs()
                    m.forward = m.forward_fuse  # update forward
                if isinstance(m, RepVGGDW):
                    m.fuse()
                    m.forward = m.forward_fuse
                if isinstance(m, v10Detect):
                    m.fuse()  # remove one2many head
            self.info(verbose=verbose)

        return self
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @xifeng0126, thank you for your interest in Ultralytics 🚀! We recommend visiting the [Docs](https://docs.ultralytics.com/) where you’ll find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions about export and model fusing are already covered there.

If this is a 🐛 Bug Report or a request related to missing functionality, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us understand your environment and code path. This assists our team in debugging and providing a solution more efficiently.

If this is a custom training or export ❓ Question, please include as much detail as possible, such as code snippets, model definitions, and the specific YOLOE variant you are working with.

Join the Ultralytics community for more help and discussion:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth Q&A: [Discourse](https://community.ultralytics.com/)
- Discussions and sharing: [Subreddit](https://www.reddit.com/r/Ultralytics/)

## Upgrade

Please ensure you’re running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to make sure your issue has not already been resolved in a recent release:

```bash
pip install -U ultralytics
```

## Environments

YOLO can run in these up-to-date verified environments (with all dependencies, [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response. An Ultralytics engineer will review your question and assist you soon 😊

**Comment by Y-T-G** [comment]
It's fused before export. You can see the fused log:

```
Ultralytics 8.3.161 🚀 Python-3.11.10 torch-2.7.0+cu126 CPU (AMD EPYC 7502 32-Core Processor)
YOLOe-v8s-seg summary (fused): 109 layers, 15,406,182 parameters, 0 gradients, 43.6 GFLOPs

PyTorch: starting from 'yoloe-v8s-seg.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 116, 8400), (1, 32, 160, 160)) (29.7 MB)

ONNX: starting export with onnx 1.17.0 opset 19...
ONNX: slimming with onnxslim 0.1.56...
ONNX: export success ✅ 3.5s, saved as 'yoloe-v8s-seg.onnx' (46.1 MB)
```

**Comment by xifeng0126** [comment]
> It's fused before export. You can see the fused log:
> 
> ```
> Ultralytics 8.3.161 🚀 Python-3.11.10 torch-2.7.0+cu126 CPU (AMD EPYC 7502 32-Core Processor)
> YOLOe-v8s-seg summary (fused): 109 layers, 15,406,182 parameters, 0 gradients, 43.6 GFLOPs
> 
> PyTorch: starting from 'yoloe-v8s-seg.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 116, 8400), (1, 32, 160, 160)) (29.7 MB)
> 
> ONNX: starting export with onnx 1.17.0 opset 19...
> ONNX: slimming with onnxslim 0.1.56...
> ONNX: export success ✅ 3.5s, saved as 'yoloe-v8s-seg.onnx' (46.1 MB)
> ```

But it does not actually remove the BN layer(the node before `/model.22/Concat`), you can try it.

![Image](https://github.com/user-attachments/assets/b95b81a4-3f51-4040-be12-38fda12c921a)

**Comment by Y-T-G** [comment]
Is that BatchNorm layer?

**Comment by Y-T-G** [comment]
There's no BN after it's fused:

```python
In [10]: model.fuse()
YOLOe-v8s-seg summary (fused): 111 layers, 15,404,318 parameters, 12,962,901 gradients, 42.9 GFLOPs

In [11]: for m in model.modules():
    ...:     if type(m) is nn.BatchNorm2d:
    ...:         print(m)
    ...:
```

**Comment by xifeng0126** [comment]
> Is that BatchNorm layer?

Sorry, I think there is something wrong with my expression, my confusion is that the forward method in BNContrastiveHead does not change after fuse, you can see the corresponding code [here](https://github.com/ultralytics/ultralytics/blob/72804697eb716a9d27e8ee8e70c09631b5adf396/ultralytics/nn/modules/block.py#L826C1-L853C1).
```
  def fuse(self):
        """Fuse the batch normalization layer in the BNContrastiveHead module."""
        del self.norm
        del self.bias
        del self.logit_scale
        self.forward = self.forward_fuse

  def forward_fuse(self, x: torch.Tensor, w: torch.Tensor) -> torch.Tensor:
        """Passes input out unchanged."""
        return x
  def forward(self, x: torch.Tensor, w: torch.Tensor) -> torch.Tensor:
        """
        Forward function of contrastive learning with batch normalization.

        Args:
            x (torch.Tensor): Image features.
            w (torch.Tensor): Text features.

        Returns:
            (torch.Tensor): Similarity scores.
        """
        x = self.norm(x)
        w = F.normalize(w, dim=-1, p=2)

        x = torch.einsum("bchw,bkc->bkhw", x, w)
        return x * self.logit_scale.exp() + self.bias
```
After fuse, there shouldn't be any more einsum calculations, right? Actually, I got the results I wanted after exporting the model using the [code](https://github.com/THU-MIG/yoloe/blob/main/ultralytics/nn/tasks.py#L201-L236) from the YOLOE repository.

![Image](https://github.com/user-attachments/assets/005afe8b-7774-439b-89f4-7363ed2e0996)

**Comment by xifeng0126** [comment]
I refine the fuse method, it works.   
```
def fuse(self, verbose=True):
        """
        Fuse the `Conv2d()` and `BatchNorm2d()` layers of the model into a single layer for improved computation
        efficiency.

        Returns:
            (torch.nn.Module): The fused model is returned.
        """
        if not self.is_fused():
            for m in self.model.modules():
                if isinstance(m, (Conv, Conv2, DWConv)) and hasattr(m, "bn"):
                    if isinstance(m, Conv2):
                        m.fuse_convs()
                    m.conv = fuse_conv_and_bn(m.conv, m.bn)  # update conv
                    delattr(m, "bn")  # remove batchnorm
                    m.forward = m.forward_fuse  # update forward
                if isinstance(m, ConvTranspose) and hasattr(m, "bn"):
                    m.conv_transpose = fuse_deconv_and_bn(m.conv_transpose, m.bn)
                    delattr(m, "bn")  # remove batchnorm
                    m.forward = m.forward_fuse  # update forward
                if isinstance(m, RepConv):
                    m.fuse_convs()
                    m.forward = m.forward_fuse  # update forward
                if isinstance(m, RepVGGDW):
                    m.fuse()
                    m.forward = m.forward_fuse
                if isinstance(m, v10Detect):
                    m.fuse()  # remove one2many head
                # YOLOE update forward
                device = next(self.model.parameters()).device
                if isinstance(m, YOLOEDetect) and hasattr(self, 'pe'):
                    assert(isinstance(self, YOLOEModel))
                    m.fuse(self.pe.to(device))
                    
            self.info(verbose=verbose)
        return self
```

**Comment by Y-T-G** [comment]
Can you open a PR for it?

**Comment by xifeng0126** [comment]
> Can you open a PR for it?

Sure，I'll do it later.

**Comment by glenn-jocher** [comment]
Great! Thank you for identifying the issue and providing a solution. When you open the PR, please make sure to include the specific case you tested and the before/after ONNX model comparison to demonstrate that the BNContrastiveHead is properly fused. The addition of YOLOEDetect handling in the fuse method looks like the right approach to ensure proper export optimization.
